honeybee_grasshopper_energy/src/HB Apply Window Construction.py

```diff
@@ -0,0 +1,120 @@
+            cardinal direction, starting with north and moving clockwise.
+    
+    Returns:
+        hb_obj: The input honeybee objects with their constructions edited.
+"""
+
+ghenv.Component.Name = "HB Apply Window Construction"
+ghenv.Component.NickName = 'ApplyWindowConstr'
+ghenv.Component.Message = '0.1.0'
+ghenv.Component.Category = 'Energy'
+ghenv.Component.SubCategory = '1 :: Constructions'
+ghenv.Component.AdditionalHelpFromDocStrings = '3'
+
+
+try:  # import the honeybee-energy extension
+    from honeybee_energy.lib.constructions import window_construction_by_name
+except ImportError as e:
+    raise ImportError('\nFailed to import honeybee_energy:\n\t{}'.format(e))
+
+try:  # import the core honeybee dependencies
+    from honeybee.boundarycondition import Outdoors
+    from honeybee.facetype import Wall
+    from honeybee.room import Room
+    from honeybee.face import Face
+    from honeybee.aperture import Aperture
+    from honeybee.door import Door
+    from honeybee.orientation import angles_from_num_orient, face_orient_index
+except ImportError as e:
+    raise ImportError('\nFailed to import honeybee:\n\t{}'.format(e))
+
+try:  # import the ladybug_rhino dependencies
+    from ladybug_rhino.grasshopper import all_required_inputs
+except ImportError as e:
+    raise ImportError('\nFailed to import ladybug_rhino:\n\t{}'.format(e))
+
+
+def is_exterior_wall(face):
+    """Check whether a given Face is an exterior Wall."""
+    return isinstance(face.boundary_condition, Outdoors) and \
+        isinstance(face.type, Wall)
+
+
+if all_required_inputs(ghenv.Component):
+    # duplicate the initial objects
+    hb_obj = [obj.duplicate() for obj in _hb_obj]
+    
+    # process the input constructions
+    for i, constr in enumerate(_constr):
+        if isinstance(constr, str):
+            _constr[i] = window_construction_by_name(constr)
+    
+    # error message for unrecognized object
+    error_msg = 'Input _hb_obj must be a Room, Face, Aperture, or Door. Not {}.'
+    
+    # assign the constructions
+    if len(_constr) == 1:  # assign indiscriminately, even if it's a horizontal object
+        for obj in hb_obj:
+            if isinstance(obj, (Aperture, Door)):
+                obj.properties.energy.construction = _constr[0]
+            elif isinstance(obj, Face):
+                for ap in obj.apertures:
+                    ap.properties.energy.construction = _constr[0]
+            elif isinstance(obj, Room):
+                for face in obj.faces:
+                    if is_exterior_wall(face):
+                        for ap in face.apertures:
+                            ap.properties.energy.construction = _constr[0]
+            else:
+                raise TypeError(error_msg.format(type(obj)))
+    else:  # assign constructions only to non-horizontal objects based on cardinal direction
+        angles = angles_from_num_orient(len(_constr))
```

Comment:
Do I understand correctly that the value that these components add over using construction-set is the fact that you can assign different constructions based on orientation? I like the components but I'm trying to understand why are they required and why we have them separated based on type.

Reply:
Yes, they assign constructions to individual Faces, Apertures, Doors, or Shades, thereby overriding the ConstructionSet.

They serve a few different purposes:
- Someone wants to customize a model that they are importing from JSON, idf, gbXML, etc. at the level of detail that they would in a face-by-face model workflow. Since they don't have access to assign individual Face or Aperture constructions when these objects are created, this component gives people the ability to make detailed edits.
- People want to use different window/wall constructions or wall constructions for different orientations (I think this case is pretty common). The "Glazing by Ratio" component currently supports this but I know that we were hoping to get rid of these extra inputs for ep_constr_ and rad_mat_ on the component. 
- These current components allow such orientation-specific assignment when the Apertures are obtained by some means other than "Apertures by Ratio", like assignment using the "Add SubFace" component or, again import from an external file.

Reply:
I think one question that stands is why there are 3 components based on type instead of one with 4 inputs.

![image](https://user-images.githubusercontent.com/2915573/70850455-73625180-1e58-11ea-977a-8e6710fb4192.png)

Also I still don't know how one would identify that the construction be set for doors and not the wall or the other way around.

Will be similar components for roof, ceiling, etc or since they don't have different directions one should use the construction-set component instead?

Reply:
I think I got my answer when checking set adiabatic component. These components are actually "set construction based on orientation". Otherwise one can use the construction-sub-set to change the construction type.